frontend/apps/reader/modules/readerdevicestatus.lua

```diff
@@ -12,114 +16,274 @@ local ReaderDeviceStatus = InputContainer:new{
-                    title_align = "center",
-                    title_face = Font:getFace("infofont"),
+                powerd:setDismissBatteryStatus(false)
+            elseif powerd:getDismissBatteryStatus() ~= true and battery_capacity <= self.battery_threshold then
+                powerd:setDismissBatteryStatus(true)
+                UIManager:show(ConfirmBox:new {
+                    text = T(_("Low battery level: %1%\n\nDismiss low battery alarm?"), battery_capacity),
+                    ok_text = _("Dismiss"),
                     dismissable = false,
-                    buttons = {
-                        {
-                            {
-                                text = _("Dismiss"),
-                                callback = function()
-                                    UIManager:close(low_battery_info)
-                                    powerd:setDissmisBatteryStatus(true)
-                                    UIManager:scheduleIn(300, self.checkLowBattery)
-                                end,
-                            },
-                        },
-                    }
-                }
-                UIManager:show(low_battery_info)
-                return
-            elseif powerd:getDissmisBatteryStatus() and battery_capacity > threshold then
-                powerd:setDissmisBatteryStatus(false)
+                    cancel_callback = function()
+                        powerd:setDismissBatteryStatus(false)
+                    end,
+                })
+            elseif powerd:getDismissBatteryStatus() and battery_capacity > self.battery_threshold then
+                powerd:setDismissBatteryStatus(false)
             end
-            UIManager:scheduleIn(300, self.checkLowBattery)
+            UIManager:scheduleIn(self.battery_interval * 60, self.checkLowBattery)
         end
-        self.ui.menu:registerToMainMenu(self)
         self:startBatteryChecker()
-    else
-        self.checkLowBattery = nil
     end
- end
+
+    self.memory_interval = G_reader_settings:readSetting("device_status_memory_interval") or 5
+    self.memory_threshold = G_reader_settings:readSetting("device_status_memory_threshold") or 100
+    self.checkHighMemory = function()
+        local statm = io.open("/proc/self/statm", "r")
+        if statm then
+            local dummy, rss = statm:read("*number", "*number")
+            statm:close()
+            rss = math.floor(rss * 4096 / 1024 / 1024)
+            if rss >= self.memory_threshold then
+                if Device:canRestart() then
+                    if G_reader_settings:isTrue("device_status_memory_auto_restart") then
+                        UIManager:show(InfoMessage:new{
+                            text = _("High memory usage!\n\nKOReader is restarting..."),
+                            icon = "notice-warning",
+                        })
+                        self.ui:handleEvent(Event:new("Restart"))
```

Comment:
Would it make sense to only do that when ReaderUI or FileManager are the top level widget ?
We may get a large increase when launching some Dict/Wikipedia lookup and reading, and doing more of them - if restarting, we're kicked out of our lookup session, and have to redo it all. May be just fallback to the ConfirmBox in this case ?
But if ReaderUI is top, and "Start with last document", a restart would feel nearly like nothing :)